cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -307,12 +310,29 @@ public void removeAdapter(Id.Namespace namespace, String adapterName)
     throws AdapterNotFoundException, CannotBeDeletedException {
 
     AdapterStatus adapterStatus = getAdapterStatus(namespace, adapterName);
+    AdapterSpecification adapterSpec = getAdapter(namespace, adapterName);
```

Comment:
Can you add a TODO. The logic is not transactional and there can be race that one thread is creating and the other thread is deleting.